moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp

```diff
@@ -0,0 +1,140 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2011, Willow Garage, Inc.
+ *  Copyright (c) 2018, Michael 'v4hn' Goerner
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Authors: Ioan Sucan, Michael Goerner */
+
+#include <tf/transform_datatypes.h>
+#include <moveit/ompl_interface/ompl_interface.h>
+#include <moveit/robot_model_loader/robot_model_loader.h>
+#include <moveit/profiler/profiler.h>
+
+#include <moveit/planning_scene_monitor/planning_scene_monitor.h>
+
+#include <moveit/kinematic_constraints/utils.h>
+
+#include <boost/math/constants/constants.hpp>
+#include <sstream>
+
+static const std::string ROBOT_DESCRIPTION = "robot_description";
+
+static const std::string CONSTRAINT_PARAMETER = "constraints";
+
+void computeDB(const planning_scene::PlanningScenePtr& scene, moveit_msgs::Constraints& constraints,
+               ros::NodeHandle& nh)
+{
+  // required by addConstraintApproximation
+  scene->getCurrentStateNonConst().update();
+
+  ompl_interface::OMPLInterface ompl_interface(scene->getRobotModel());
+
+  ompl_interface::ConstraintApproximationConstructionOptions opt;
+
+  // number of states in joint space approximation
+  opt.samples = nh.param<int>("state_cnt", 10000);
+
+  // useful default parameterization
```

Comment:
Seems like making a class and putting all the parameter reading into a separate method would be cleaner